text/0000-snapshot-system.md

```diff
@@ -0,0 +1,207 @@
+
+# Summary
+[summary]: #summary
+
+The Snapshot System will store state (NetworkVariable updates) and send them in blocks (called Snapshot Messages) at regular intervals (called ticks). The Snapshot System will also store events (Spawn commands, for example) and deliver them reliably.
+
+This RFC covers a first usage of the Snapshot System for the first release. This first usage is for sending a first type of event: Spawn commands. Despawns commands will be handled the same way, but to keep the text easy to read, we'll write `spawn` to refer to either type.
+
+Throughout this RFC, when we refer to time, we refer to Network Ticks. Those are described in the NetworkTime RFC
+
+# Motivation
+[motivation]: #motivation
+
+The motivation for the Snapshot System are:
+
+- Reducing and controlling bandwidth usage
+- Leveraging unreliable channels
+- Being more tolerant to packet loss
+- Allowing good responsiveness even when there’s packet loss
+- Interfacing with AoI
+- Providing a foundation for future systems like rollback and replay
+- Offering consistency between different features (you don’t update an object that is not spawned)
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+### The Snapshot System
+
+The `NetworkManager` will expose a `SnapshotSystem` via a public getter. Under normal operation, the existence Snapshot System should have no direct impact on game code. It should improve delivery and performance, but game code should not directly interact with it.
+
+### The Snapshot data structure
+
+Each machine will have a snapshot that captures, in-memory, the knowledge it has of various NetworkVariables values at specific ticks, and objects that were spawned. As shown below, a Snapshot will contain:
+- NetworkVariables values at specific ticks
+- Spawn commands, with all the spawn parameters, the tick and a bitmask of target connections 
+
+![](0000-snapshot-system/snapshot.rfc.png)
+
+### The Snapshot Message 
+
+Machines will exchange Snapshot Messages. A Snapshot Message is nothing but a partial Snapshot (plus a header). Those messages will be used to synchronize the state and replicate events. Apart from containing less elements than Snapshots, the only difference between Snapshot Messages and Snapshot is that Snapshot messages don't need the bit mask for destination.
+
+Additionally, a Snapshot Message will include a header that contains:
+- a sequence number unique to this message
+- the tick it was sent on
+- acknowledgment for previous messages from this connection (in the form of a bitmask over the _n_ previous sequence numbers)
+
+![](0000-snapshot-system/snapshot.message.rfc.png)
+
+### Eventual Consistency
+
+The model we will use for synchronisation is eventual consistency. Any Snapshot Message might be lost, but retransmits(1) will have the information eventually arrive to its destination. In the case of packet loss, some variables still get to update.
```

Comment:
I'm still not sold on this but fine with going forward. I think full state /delta snapshots are better for co-op and other small scale games.